src/shogun/kernel/GaussianKernel.cpp

```diff
@@ -90,19 +91,27 @@ void CGaussianKernel::set_width(float64_t w)
 
 SGMatrix<float64_t> CGaussianKernel::get_parameter_gradient(const TParameter* param, index_t index)
 {
+	using std::exp;
+	using std::log;
+
 	require(lhs, "Left hand side features must be set!");
-	require(rhs, "Rightt hand side features must be set!");
+	require(rhs, "Right hand side features must be set!");
 
 	if (!strcmp(param->m_name, "log_width"))
 	{
 		SGMatrix<float64_t> derivative=SGMatrix<float64_t>(num_lhs, num_rhs);
+		stan::math::var log_width = m_log_width;	
+		auto constant_part = exp(log_width * 2.0) * 2.0;
+		
 		for (int k=0; k<num_rhs; k++)
 		{
 #pragma omp parallel for
 			for (int j=0; j<num_lhs; j++)
 			{
-				float64_t element=distance(j, k);
-				derivative(j, k) = std::exp(-element) * element * 2.0;
+				auto f = exp(-CShiftInvariantKernel::distance(j, k) / constant_part);
+				f.grad();
```

Comment:
Wow it is quite simple to do this for the bandwidth ... I guess I tried to diff wrt the features last time I tried, which raised problems regarding passing the matrices to stan...but this is cool!

I assume the same results as there are tests covering this?
what are the computational differences compared to the previous line? Maybe a benchmark would be good?

Reply:
yes, the results are correct and are test in unittests, mostly by GP stuff.
The computation is very slow... Basically it creates a new variable in the  heap in the inner loop which is in the heap, and then it redoes the gradient calculation. 
What we need is a way to calculate the "formula" of the gradient once outside the loop and then reuse it.. Not sure how to do this in stan or if it is even possible